### DISCUSSION TITLE
Support for OAuth2 ?

### DISCUSSION THREAD

**Discussion by StefanArcelor** [testingisdocumenting/webtau#1297] [Q&A]
The API I would like to test requires an OAuth2 authentication with a Bearer Token and an Auth URL. Is this already supported ?

**Comment by MykolaGolubyev** [comment]
Thank you for the question, @StefanArcelor.
WebTau supports OAuth2. You can provide `Authorization` field with value `Bearer <token>` as a header.
There is support for explicit and implicit headers. 

*explicit header*
```
http.get("/end-point", http.header("Authorization", "Bearer $token")) {
    // assertions go here
}
```

With implicit header all you calls will automatically have auth header injected.
https://testingisdocumenting.org/webtau/HTTP/header#implicit-header

Additionally you can use Persona auth to streamline authorization testing. Your implicit header provider can use information about current persona and inject different tokens based on that
https://testingisdocumenting.org/webtau/persona/HTTP-persona

```
scenario("my bank balance") {
    Alice {
        http.get("/statement") {
            balance.shouldBe > 100
        }
    }

    Bob {
        http.get("/statement") {
            balance.shouldBe < 50
        }
    }
}
```

Let me know if you have questions on how to get the actual token

**Reply by StefanArcelor** [reply]
Based on the documentation, I've tried with underneath code. How can I specify in this code Grant Type, Auth URL, Client ID and Scope, like I do in Postman (see https://i.imgur.com/2V3xlYp.png) ?

`class Auth {
    static HttpHeader authHeader(String fullUrl, String url, HttpHeader original) {
        def token = generateToken()
        return original.with([Authorization: "Bearer $token"])
    }

    private static String generateToken() {
        return "jwt-token"
    }
}

scenario('Export Production Safeties data') {
    httpHeaderProvider = Auth.&authHeader
`

**Reply by MykolaGolubyev** [reply]
Any chance you can elaborate to me steps you took to acquire the token using Postman?

**Reply by MykolaGolubyev** [reply]
@StefanArcelor I am working on an article that demos how to do Okta Auth flow through clientId and clientSecret pair to acquire and use auth token. Let me know if this is something that can be helpful for you.

**Comment by StefanArcelor** [comment]
In Postman I'm using the Authorization info like this:

- Type: "OAuth2.0"
- Add Auth Data to "Request Headers"
- Grant Type: "Implicit"
- Auth URL: "my auth url"
- client id: "my client id"
- scope: "SwaggerApi"
- Client Authentication: "Send as Basic Auth header"
- Get New Access Token

**Reply by MykolaGolubyev** [reply]
thanks for sharing the steps.
As a starting point, can we try a WebTau script that does an HTTP call with explicit header set to Bearer and token. 
```
def token = "your hardcoded token you receive from postman"
http.get("<yoururl>", http.header("Authorization", "Bearer $token"))
```

want to make sure that basics work.

**Reply by MykolaGolubyev** [reply]
Also feel free to join discord server to have more convenient means to ask questions. I just started the server: https://discord.gg/Z5yjatDy
We can also discuss things with voice if it is easier

**Reply by MykolaGolubyev** [reply]
@StefanArcelor did you have a chance to try a basic test?

**Reply by MykolaGolubyev** [reply]
@StefanArcelor checking just in case
